src/Features/Core/Shared/Utilities/PatternMatcher.cs

```diff
@@ -19,19 +19,129 @@ namespace Microsoft.CodeAnalysis.Shared.Utilities
     /// </summary>
     internal sealed class PatternMatcher
     {
+        // First we break up the pattern given by dots.  Each portion of the pattern between the
+        // dots is a 'Segment'.  The 'Segment' contains information about the entire section of 
+        // text between the dots, as well as information about any individual 'Words' that we 
+        // can break the segment into.
+        private struct Segment
+        {
+            // Information about the entire piece of text between the dots.  For example, if the 
+            // text between the dots is 'GetKeyword', then TotalTextChunk.Text will be 'GetKeyword' and 
+            // TotalTextChunk.CharacterSpans will correspond to 'G', 'et', 'K' and 'eyword'.
+            public readonly TextChunk TotalTextChunk;
+
+            // Information about the subwords compromising the total word.  For example, if the 
+            // text between the dots is 'GetKeyword', then the subwords will be 'Get' and 'Keyword'
+            // Those individual words will have CharacterSpans of ('G' and 'et') and ('K' and 'eyword')
+            // respectively.
+            public readonly TextChunk[] SubWordTextChunks;
+
+            public Segment(string text, bool verbatimIdentifierPrefixIsWordCharacter)
+            {
+                this.TotalTextChunk = new TextChunk(text);
+                this.SubWordTextChunks = BreakPatternIntoTextChunks(text, verbatimIdentifierPrefixIsWordCharacter);
+            }
+
+            public bool IsInvalid => this.SubWordTextChunks.Length == 0;
+
+            private static int CountTextChunks(string pattern, bool verbatimIdentifierPrefixIsWordCharacter)
```

Comment:
The Count and Break function are simply moves from below.  Unfortunately, github doesn't show moves well at all.